deployments/cluster-resources.nix

```diff
@@ -1,132 +1,47 @@
-    instanceTenancy = "default";
-    enableDnsSupport = true;
-    enableDnsHostnames = true;
-    cidrBlock = "10.0.0.0/16";
-  };
-
-  vpcSubnets.cluster-subnet =
-    { resources, lib, ... }:
-    {
-      inherit region zone;
-      vpcId = resources.vpc.cluster-vpc;
-      cidrBlock = "10.0.44.0/24";
-      mapPublicIpOnLaunch = true;
-    };
-
-  elasticIPs = if production then
-    {
-      balancer-ip = mkIP;
-      witness1-ip = mkIP;
-      witness2-ip = mkIP;
-      witness3-ip = mkIP;
-      educator-ip = mkIP;
-    } else {};
-
-  ec2SecurityGroups.cluster-http-public-sg = mkSG [
-    { fromPort =  80; toPort =  80; sourceIp = "0.0.0.0/0"; }
-    { fromPort = 443; toPort = 443; sourceIp = "0.0.0.0/0"; }
-  ];
-
-  ec2SecurityGroups.cluster-ssh-private-sg = mkSG [
-    { fromPort =    22; toPort =    22; sourceIp = vpc.cluster-vpc.cidrBlock; }
-  ];
-
-  ec2SecurityGroups.cluster-ssh-public-sg = mkSG [
-    { fromPort =    22; toPort =    22; sourceIp = "0.0.0.0/0"; }
-  ];
-
-  ec2SecurityGroups.cluster-witness-public-sg = mkSG [
-    { fromPort =  4010; toPort =  4011; sourceIp = "0.0.0.0/0"; }
-  ];
-
-  ec2SecurityGroups.cluster-witness-private-sg = mkSG [
-    { fromPort =  4010; toPort =  4011; sourceIp = vpc.cluster-vpc.cidrBlock; }
-  ];
-
-  ec2SecurityGroups.cluster-witness-api-public-sg = mkSG [
-    { fromPort =  4030; toPort =  4030; sourceIp = "0.0.0.0/0"; }
-  ];
-
-  ec2SecurityGroups.cluster-witness-api-private-sg = mkSG [
-    { fromPort =  4030; toPort =  4030; sourceIp = vpc.cluster-vpc.cidrBlock; }
-  ];
-
-  ec2SecurityGroups.cluster-educator-api-private-sg = mkSG [
-    { fromPort =  4040; toPort =  4040; sourceIp = vpc.cluster-vpc.cidrBlock; }
-  ];
-
-  vpcRouteTables.cluster-route-table =
-    { resources, ... }:
-    {
-      inherit region;
-      vpcId = resources.vpc.cluster-vpc;
-    };
-
-  vpcRouteTableAssociations.cluster-assoc =
-    { resources, ... }:
-    {
-      inherit region;
-      subnetId = resources.vpcSubnets.cluster-subnet;
-      routeTableId = resources.vpcRouteTables.cluster-route-table;
-    };
-
-  vpcInternetGateways.cluster-igw =
-    { resources, ... }:
-    {
-      inherit region;
-      vpcId = resources.vpc.cluster-vpc;
-    };
-
-  vpcRoutes.cluster-route =
-    { resources, ... }:
-    {
-      inherit region;
-      routeTableId = resources.vpcRouteTables.cluster-route-table;
-      destinationCidrBlock = "0.0.0.0/0";
-      gatewayId = resources.vpcInternetGateways.cluster-igw;
-    };
+  lib = import ./lib.nix { inherit region vpcCidr; };
+  inherit (lib) withVPC sg dns publicSubnet;
+in
+  {
+    resources = {
+      ec2KeyPairs.cluster-keypair = { inherit region; };
+
+      elasticIPs.balancer-eip = { inherit region; vpc = true; };
+
+      vpcSubnets.a-subnet = publicSubnet vpcId "${region}a" "10.1.41.0/24";
+      vpcSubnets.b-subnet = publicSubnet vpcId "${region}b" "10.1.42.0/24";
+      vpcSubnets.c-subnet = publicSubnet vpcId "${region}c" "10.1.43.0/24";
```

Comment:
Not sure I got it. `region` is the region code, apparently one from the list: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-available-regions

However, you append `a`, `b` and `c` to this code and that becomes a region identifier for a subnet. But there is no valid region identifier `eu-west-2a`. What do I miss? Where do actual region identifiers emerge? From what I see, different witnesses just use different subnets defined here, but I don't see how those subnets relate to different availability zones recognizable by Amazon.

Reply:
The region `eu-west-2` has three availability zones: `eu-west-2{a,b,c}`. A VPC can span an entire region, but subnets can not span multiple AZ, so there needs to be one in each, and they need to be linked via route table so that nodes can talk across AZ.

The basic idea is to distribute the network across AZ for better reliability.

Reply:
@mkaito Got it, but didn't we want to spread the cluster around several _regions_, in order to e. g. avoid problems of connecting to the cluster from Asia?